### PR TITLE
Task refactor units

### PR DESCRIPTION
**Pull request by royfalk**
Another PR that is part of #73. This one removes the generic unit subclasses (missile_generic.h for example) and aligns subclasses around the file name (class Missile).

Code Changes:
- [Yes ] Have the PR Validation Tests been run? 

Issues:
- Fixed a bug in the collision code. It seems an Asteroid unit is the entire field and actual asteroids are subunits. When colliding with one, I called the kill, which caused the entire field to disappear.

Purpose:
- What release is this for? 0.7 probably.
- Is there a project or milestone we should apply this to?

### CODE REVIEW

File: engine/src/cmd/building.h [hunk 1]
```diff
@@ -2,46 +2,79 @@
...
+              bool SubUnit,
+              int faction,
+              const std::string &unitModifications = std::string( "" ),
+              Flightgroup *fg = NULL );
 
     friend class UnitFactory;
 
 public:
 
+    virtual enum clsptr isUnit() const
```
Comment (BenjamenMeyer) [inline engine/src/cmd/building.h:46]: we probably should rename the `clsptr` enum to be more descriptive

File: engine/src/cmd/nebula.cpp [hunk 2]
```diff
@@ -9,15 +9,147 @@
...
+}
+
+
+
+void Nebula::beginElement( void *Userdata, const XML_Char *name, const XML_Char **atts )
+{
+    ( (Nebula*) Userdata )->beginElem( std::string( name ), AttributeList( atts ) );
+}
+
+void Nebula::beginElem( const std::string &name, const AttributeList &atts )
```
Comment (BenjamenMeyer) [inline engine/src/cmd/nebula.cpp:61]: this probably should also be called `beginElement()` and allow for function overloading to do the work.

**Review by BenjamenMeyer — COMMENTED** [review]
Overall I don't see anything in particularly wrong; left a couple comments on further refactoring - this PR or another. Let's get this play tested and it should be good to merge.

**Comment by evertvorster** [comment]
Hi there! I have not fully tested this PR, but I am noticing something weird. 

Top speed with SPEC is only 10 x C, and not 100 x C that I am used to getting. 
I have noticed it on another of Roy's branches, too, but thought I was mistaken. It's only now that I have switched back and forth between master and this branch that I know for certain!

**Comment by stephengtuggy** [comment]
Well, we'll certainly need to fix the SPEC top speed issue before merging this.

I will still try to review this PR in detail today or tomorrow.

**Review by stephengtuggy — CHANGES_REQUESTED** [review]
Mostly looks good. Just a couple of things I noticed.

File: engine/src/cmd/nebula.h [hunk 3]
```diff
@@ -20,17 +49,65 @@ class GameNebula : public GameUnit< Nebula >
...
-    GameNebula( const Nebula& );
-/// assignment operator forbidden
-    GameNebula& operator=( const Nebula& );
+protected:
+    /// default constructor forbidden
+        Nebula();
+    /// copy constructor forbidden
+        Nebula( const Nebula& );
+    /// assignment operator forbidden
+        Nebula& operator=( const Nebula& );
```
[the remark this answers is not in the document]
Comment (stephengtuggy) [reply engine/src/cmd/nebula.h:61]: @royfalk Update: It looks like the optimum syntax to use in C++11 is `= delete`. Any objections to using that form?

File: engine/src/cmd/missile.cpp [hunk 4]
```diff
@@ -178,15 +178,15 @@ float Missile::ExplosionRadius()
...
 
 
-enum clsptr Missile::isUnit() const
-{
-    return MISSILEPTR;
-}
+//enum clsptr Missile::isUnit() const
+//{
+//    return MISSILEPTR;
+//}
```
[the remark this answers is not in the document]
Comment (stephengtuggy) [reply engine/src/cmd/missile.cpp:184]: I guess this isn't used anywhere, or there would have been a compile error. So I suppose it's OK to leave out. I was just curious.

File: engine/src/cmd/nebula.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (stephengtuggy) [reply engine/src/cmd/nebula.h:61]: Roy, was `= delete` the syntax you wanted to use originally for forbidden constructors? 😆 If so, you were smarter than me or @BenjamenMeyer in this case. We weren't familiar with the C++11 syntax yet lol

[the remark this answers is not in the document]
Comment (stephengtuggy) [reply engine/src/cmd/nebula.h:61]: And the more I think about it, the more I realize that this isn't really a blocker. If you want to merge master into this branch, we can probably go ahead and merge the PR. Any constructor cleanup can be done later.

**Review by stephengtuggy — APPROVED** [review]
After further consideration, this is fine for now. I think we can go ahead and merge it.

**Comment by stephengtuggy** [comment]
It looks like the top speed 9C issue had made it into master already, before this PR. So I think this PR may be safe to merge. I'll just want to do a little play testing on it myself, to double check.

**Comment by stephengtuggy** [comment]
After some play testing, this looks good to me. Top speed is back to 97C. 👍🏼 Merging.

**Pull request merged**